server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -588,6 +594,33 @@ private static void parseNonDynamicArray(DocumentParserContext context, final St
                 parseValue(context, lastFieldName);
             }
         }
+        postProcessDynamicArrayMapping(context, lastFieldName);
+    }
+
+    /**
+     * Arrays that have been classified as floats and meet specific criteria are re-mapped to dense_vector.
+     */
+    private static void postProcessDynamicArrayMapping(DocumentParserContext context, String fieldName) {
+        if (context.indexSettings().getIndexVersionCreated().onOrAfter(DYNAMICALLY_MAP_DENSE_VECTORS_INDEX_VERSION)) {
+            final MapperBuilderContext builderContext = context.createDynamicMapperBuilderContext();
+            final String fullFieldName = builderContext.buildFullName(fieldName);
+            final List<Mapper> mappers = context.getDynamicMappers().get(fullFieldName);
+            if (mappers == null
+                || context.isFieldAppliedFromTemplate(fullFieldName)
+                || context.isCopyToField(fullFieldName)
+                || mappers.size() < MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING
+                || mappers.size() > MAX_DIMS_COUNT
+                || mappers.stream().allMatch(m -> m instanceof NumberFieldMapper && "float".equals(m.typeName())) == false) {
+                return;
+            }
+
+            DenseVectorFieldMapper.Builder builder = new DenseVectorFieldMapper.Builder(
```

Comment:
I'm running into an issue here with subobjects, where we're hanging and timing out somewhere. This is hard to debug but I suspect it's due to the fact that we have child objects, and we might need to adjust the mapper that we create to have a new parent. More guidance here would be greatly appreciated though.

Reply:
To debug, I would add a test to [DynamicMappingTests.java](https://github.com/elastic/elasticsearch/pull/98512/files#diff-804cd3ad776a31821d8bdbe72b83e278ee1da2fd6d1bf7da753f31cd435af08c) and debug in intellij. I can help with this to see whats up.

Reply:
This test passes:
```
    public void testDefaultDenseVectorMappingsObject() throws IOException {
        DocumentMapper mapper = createDocumentMapper(topMapping(b -> b.field("numeric_detection", true)));
        BytesReference source = BytesReference.bytes(
            XContentFactory.jsonBuilder().startObject()
                .startObject("parent_object")
                .field("mapsToFloatTooSmall", Randomness.get().doubles(MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING - 1, 0.0, 5.0).toArray())
                .field("mapsToFloatTooBig", Randomness.get().doubles(MAX_DIMS_COUNT + 1, 0.0, 5.0).toArray())
                .field("mapsToDenseVector", Randomness.get().doubles(MIN_DIMS_FOR_DYNAMIC_FLOAT_MAPPING, 0.0, 5.0).toArray())
                .endObject()
                .endObject()
        );
        ParsedDocument parsedDocument = mapper.parse(new SourceToParse("id", source, XContentType.JSON));
        Mapping update = parsedDocument.dynamicMappingsUpdate();
        assertNotNull(update);
        ObjectMapper parent = (ObjectMapper) update.getRoot().getMapper("parent_object");
        assertThat(((FieldMapper) parent.getMapper("mapsToDenseVector")).fieldType().typeName(), equalTo("dense_vector"));
    }
```

Reply:
Thanks. This test also works for me (and is a good unit test to add, thank you!) but what fails is: 

- The yaml test that updates a subobject
- Running the command locally also results in a timeout waiting for a dynamic mapping update. 

The mapping does get set under the hood, it's just the index call that returns a 500 timeout. 

This tells me that generating the mappings themselves are fine, but something's not quite right down the line when we're indexing the document. The error trace doesn't have anything useful, just that a timeout happened. 

With all this I'm fairly confident that I'm creating a dynamic mapper, but something's not quite right with this mapper vs. the mapper that I'm inserting, leading to hanging somewhere. I'm still digging in how to troubleshoot it.

Reply:
Progress: I've narrowed this down to happening when dynamically computing the number of dims in a sub_object, regardless of whether it was a dynamic float or specifically mapped as dense_vector with no dims specified.